### PR TITLE
Initial restructure to add bugsnag-spring

### PR DESCRIPTION
**Pull request by simonbowring**
## Goal

In order to allow better support for spring, added a `bugsnag-spring` module. The end goal is to produce a bugsnag-spring JAR as well as the current one.

## Design

The `/src` directly was moved into its own module `bugsnag` in order to allow multiple modules. The Gradle build has been modified to build both modules. 

## Review

For the submitter, initial self-review:

- [ ] Commented on code changes inline explain the reasoning behind the approach
- [x] Reviewed the test cases added for completeness and possible points for discussion
- [ ] A changelog entry was added for the goal of this pull request
- [x] Check the scope of the changeset - is everything in the diff required for the pull request?
- This pull request is ready for:
  - [x] Initial review of the intended approach, not yet feature complete
  - [ ] Structural review of the classes, functions, and properties modified
  - [ ] Final review

For the pull request reviewer(s), this changeset has been reviewed for:

- [ ] Consistency across platforms for structures or concepts added or modified
- [ ] Consistency between the changeset and the goal stated above
- [ ] Internal consistency with the rest of the library - is there any overlap between existing interfaces and any which have been added?
- [ ] Usage friction - is the proposed change in usage cumbersome or complicated?
- [ ] Performance and complexity - are there any cases of unexpected O(n^3) when iterating, recursing, flat mapping, etc?
- [ ] Concurrency concerns - if components are accessed asynchronously, what issues will arise
- [ ] Thoroughness of added tests and any missing edge cases
- [ ] Idiomatic use of the language

### CODE REVIEW

File: examples/spring/build.gradle [hunk 1]
```diff
@@ -19,5 +19,5 @@ repositories {
 dependencies {
     compile("org.springframework.boot:spring-boot-starter")
 
-    compile rootProject
+    compile project(":bugsnag-java")
```
Comment (fractalwrench) [inline examples/spring/build.gradle]: Assuming at a later date we would switch to using bugsnag-spring here

File: bugsnag-spring/build.gradle [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+test {
+    testLogging {
+        events "passed", "skipped", "failed", "standardOut", "standardError"
+    }
+}
+
+checkstyle {
+    toolVersion = "6.16"
+    configFile = new File(rootDir, "config/checkstyle/checkstyle.xml")
+}
```
Comment (fractalwrench) [inline bugsnag-spring/build.gradle]: It may be worth extracting some of this configuration to a [separate script](https://docs.gradle.org/current/userguide/plugins.html#sec:script_plugins), then applying it to both modules. That way it would only be necessary to update checkstyle/test/source compat in one place.

File: bugsnag-spring/build.gradle [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+sourceCompatibility = 1.6
+targetCompatibility = 1.6
+
+repositories {
+    jcenter()
+}
+
+dependencies {
+}
```
Comment (fractalwrench) [inline bugsnag-spring/build.gradle]: I think there are two things we might want to support here as a dependency:

1. the local version of the 'bugsnag-java' module
2. a mavenCentral artefact of bugsnag-java

I suppose the key question is which of those to use when we're making a release of bugsnag-spring. I think the advantages of using a maven artefact would be there's no potential for bugsnag-java to get out of date with bugsnag-spring fixes. Although that would mean we always have to release bugsnag-java first, and wait for it to propagate.

What were your thoughts on this?

File: build.gradle [hunk 4]
```diff
@@ -16,31 +14,3 @@ buildscript {
 if (project.hasProperty('releasing')) {
     apply from: "release.gradle"
 }
```
Comment (fractalwrench) [inline build.gradle]: I think one approach here would be to apply `release.gradle` in both modules, and release each module as a separate artefact (with bugsnag-spring depending on bugsnag-java). It should be possible to update the script so that any module-specific values (e.g. package name) are supplied via `gradle.properties` or some equivalent mechanism.

In terms of testing this out I'd say we could mitigate the risk by:

- Triggering the task to generate a POM and manually inspecting it
- Uploading an alpha for both artefacts to JCenter/MavenCentral

Did you have this sort of approach in mind or something different?

File: examples/spring/build.gradle (hunk 1 above)
Comment (simonbowring) [reply examples/spring/build.gradle]: Correct

File: bugsnag-spring/build.gradle (hunk 3 above)
Comment (simonbowring) [reply bugsnag-spring/build.gradle]: My personal preference would be to have the local version of the 'bugsnag-java' module as a dependency so that we don't have to wait. It would also make development of the bugsnag-spring module more straightforward when changes also need to be made to bugsnag-java.
You are right about bugsnag-java potentially getting out of date. I'm not sure of the best way around this.

File: build.gradle (hunk 4 above)
Comment (simonbowring) [reply build.gradle]: This was the approach that I had in mind.

File: bugsnag-spring/build.gradle (hunk 3 above)
Comment (fractalwrench) [reply bugsnag-spring/build.gradle]: Another thing - if a project had a dependency on both bugsnag-java and bugsnag-spring, then this approach could potentially lead to some nasty issues with the classloader loading subtly different class definitions. I think this could be possible if someone misread our upgrade instructions, or if a Spring project depends on some 3rd party library that depends on bugsnag-java. I'm not sure that this scenario would be common enough to swing the argument one way or another.

We may be able to solve bugsnag-java getting out of date via tooling (e.g. checking the latest commit hash on master via a gradle plugin, before releasing bugsnag-spring).

Comment (simonbowring) [reply bugsnag-spring/build.gradle]: How long does it usually take maven releases to propagate?

Comment (fractalwrench) [reply bugsnag-spring/build.gradle]: 1-2 hours is typical, it can take longer however

Comment (simonbowring) [reply bugsnag-spring/build.gradle]: What about something like this as a compromise:
```
dependencies {
    if (project.hasProperty('releasing')) {
        compile 'com.bugsnag:bugsnag:3.+'
    } else {
        compile project(':bugsnag-java')
    }
}
```

Comment (simonbowring) [reply bugsnag-spring/build.gradle]: To keep consistent with the NDK rewrite approach, we are going to go with `compile project(':bugsnag-java')`.

File: .travis.yml [hunk 5]
```diff
@@ -7,22 +7,22 @@ cache:
...
   - $HOME/.gradle/wrapper/
-install: ./gradlew :assemble
+install: ./gradlew :bugsnag:assemble :bugsnag-spring:assemble
 
 jobs:
   include:
   - stage: test
     jdk: openjdk8
-    script: ./gradlew check :test
+    script: ./gradlew check :bugsnag:test :bugsnag-spring:test
```
Comment (fractalwrench) [inline .travis.yml]: this could be simplified to `./gradlew check test`

File: bugsnag-spring/build.gradle [hunk 6]
```diff
@@ -0,0 +1,10 @@
+apply plugin: 'java-library'
+
+apply from: '../common.gradle'
+
+repositories {
+    jcenter()
+}
+
+dependencies {
```
Comment (fractalwrench) [inline bugsnag-spring/build.gradle]: it'd be good to add a dependency on the `bugsnag` module here
